src/v/cluster/rm_stm.h

```diff
@@ -70,13 +69,45 @@ class rm_stm final : public persisted_stm {
         model::producer_identity pid;
     };
 
+    struct seq_cache_entry {
+        int32_t seq{-1};
+        model::offset offset;
+    };
+
     struct seq_entry {
+        static const int seq_cache_size = 5;
         model::producer_identity pid;
-        int32_t seq;
+        int32_t seq{-1};
+        model::offset last_offset{-1};
+        std::deque<seq_cache_entry> seq_cache;
```

Comment:
`seq_cache` holds at most five elements so I'm not sure if using `ss::circular_buffer` is significantly more efficient here. I've made an attempt to drop and replace deque with circular_buffer but the latter has `operator=` removed, it removes it from seq_entry and as a result I need to increase the scope of this PR. I don't think the change is worth the hassle, do you?

Reply:
there was quite a bit of unnecessary copying, so most of the needs for `operator=` could be fixed by moving instead of copying. the one place where copying was necessary you can use a custom `copy()` method to (1) avoid breaking reflection and (2) letting the compiler force you to decide when you want to incur the cost of copying.

this compiles for me:

```diff
diff --git a/src/v/cluster/rm_stm.cc b/src/v/cluster/rm_stm.cc
index aa9da770b..be1816da9 100644
--- a/src/v/cluster/rm_stm.cc
+++ b/src/v/cluster/rm_stm.cc
@@ -1302,7 +1302,7 @@ rm_stm::apply_snapshot(stm_snapshot_header hdr, iobuf&& tx_ss_buf) {
               .seq = seq_v0.seq,
               .last_offset = model::offset{-1},
               .last_write_timestamp = seq_v0.last_write_timestamp};
-            data.seqs.push_back(seq);
+            data.seqs.push_back(std::move(seq));
         }
     } else {
         vassert(
@@ -1328,9 +1328,13 @@ rm_stm::apply_snapshot(stm_snapshot_header hdr, iobuf&& tx_ss_buf) {
       std::make_move_iterator(data.abort_indexes.begin()),
       std::make_move_iterator(data.abort_indexes.end()));
     for (auto& entry : data.seqs) {
-        auto [seq_it, _] = _log_state.seq_table.try_emplace(entry.pid, entry);
-        if (seq_it->second.seq < entry.seq) {
-            seq_it->second = entry;
+        auto [seq_it, inserted] = _log_state.seq_table.try_emplace(
+          entry.pid, std::move(entry));
+        // try_emplace does not move from r-value reference if the insertion
+        // didn't take place so the clang-tidy warning is a false positive
+        // NOLINTNEXTLINE(hicpp-invalid-access-moved)
+        if (!inserted && seq_it->second.seq < entry.seq) {
+            seq_it->second = std::move(entry);
         }
     }
 
@@ -1394,13 +1398,13 @@ ss::future<stm_snapshot> rm_stm::take_snapshot() {
     for (auto& entry : _log_state.abort_indexes) {
         tx_ss.abort_indexes.push_back(entry);
     }
-    for (auto& entry : _log_state.seq_table) {
-        tx_ss.seqs.push_back(entry.second);
+    for (const auto& entry : _log_state.seq_table) {
+        tx_ss.seqs.push_back(entry.second.copy());
     }
     tx_ss.offset = _insync_offset;
 
     iobuf tx_ss_buf;
-    reflection::adl<tx_snapshot>{}.to(tx_ss_buf, tx_ss);
+    reflection::adl<tx_snapshot>{}.to(tx_ss_buf, std::move(tx_ss));
 
     co_return stm_snapshot::create(
       tx_snapshot::version, _insync_offset, std::move(tx_ss_buf));
diff --git a/src/v/cluster/rm_stm.h b/src/v/cluster/rm_stm.h
index 352422768..b1b2a5d21 100644
--- a/src/v/cluster/rm_stm.h
+++ b/src/v/cluster/rm_stm.h
@@ -75,13 +75,28 @@ public:
     };
 
     struct seq_entry {
-        static const int seq_cache_size = 5;
+        static constexpr size_t seq_cache_size = 5;
+
         model::producer_identity pid;
         int32_t seq{-1};
         model::offset last_offset{-1};
-        std::deque<seq_cache_entry> seq_cache;
+        ss::circular_buffer<seq_cache_entry> seq_cache;
         model::timestamp::type last_write_timestamp;
 
+        seq_entry copy() const {
+            seq_entry ret;
+            ret.pid = pid;
+            ret.seq = seq;
+            ret.last_offset = last_offset;
+            ret.seq_cache.reserve(seq_cache.size());
+            std::copy(
+              seq_cache.cbegin(),
+              seq_cache.cend(),
+              std::back_inserter(ret.seq_cache));
+            ret.last_write_timestamp = last_write_timestamp;
+            return ret;
+        }
+
         void update(int32_t new_seq, model::offset new_offset) {
             if (new_seq < seq) {
                 return;
```

Reply:
```
On the other hand, deques typically have large minimal memory cost; a deque holding just one element has to allocate its full internal array (e.g. 8 times the object size on 64-bit libstdc++; 16 times the object size or 4096 bytes, whichever is larger, on 64-bit libc++).
```

Reply:
That is from cppreference regarding std::deque. So for us, according to that, it's 4K allocation

Reply:
An interesting thing to note is that libc++ keeps the old block for reuse, whilst libstdc++ doesn't.

Sauce: https://www.reddit.com/r/cpp/comments/r3obmr/the_libc_implementation_of_the_stl_deque_push/hmdtl6i/

Reply:
cray. std:: really does have a lot of bits and pieces that aren't well suited for Seastar apps. Hail std::vector